pkg/query/v2/logical/analyzer.go

```diff
@@ -0,0 +1,262 @@
+	metadataSchema "github.com/apache/skywalking-banyandb/banyand/metadata/schema"
+	"github.com/apache/skywalking-banyandb/banyand/tsdb"
+	"github.com/apache/skywalking-banyandb/pkg/convert"
+)
+
+var (
+	ErrFieldNotDefined            = errors.New("field is not defined")
+	ErrInvalidConditionType       = errors.New("invalid pair type")
+	ErrIncompatibleQueryCondition = errors.New("incompatible query condition type")
+	ErrInvalidSchema              = errors.New("invalid schema")
+	ErrIndexNotDefined            = errors.New("index is not define for the field")
+	ErrTraceIDWrongType           = errors.New("trace id type should be string")
+	ErrStateWrongType             = errors.New("state type should be int")
+)
+
+var (
+	DefaultLimit uint32 = 20
+)
+
+type Tag struct {
+	familyName, name string
+}
+
+func NewTag(family, name string) *Tag {
+	return &Tag{
+		familyName: family,
+		name:       name,
+	}
+}
+
+// NewTags create an array of Tag within a TagFamily
+func NewTags(family string, tagNames ...string) []*Tag {
+	tags := make([]*Tag, len(tagNames))
+	for i, name := range tagNames {
+		tags[i] = NewTag(family, name)
+	}
+	return tags
+}
+
+// GetCompoundName is only used for error message
+func (t *Tag) GetCompoundName() string {
+	return t.familyName + "$" + t.name
+}
+
+func (t *Tag) GetTagName() string {
+	return t.name
+}
+
+func (t *Tag) GetFamilyName() string {
+	return t.familyName
+}
+
+type Analyzer struct {
+	indexRuleRepo        metadataSchema.IndexRule
+	indexRuleBindingRepo metadataSchema.IndexRuleBinding
+	metadataRepoImpl     metadata.Repo
+}
+
+func DefaultAnalyzer() *Analyzer {
+	indexRule, _ := metadataSchema.NewIndexRule()
+	indexRuleBinding, _ := metadataSchema.NewIndexRuleBinding()
+	metadataService, _ := metadata.NewService(context.TODO())
+
+	return &Analyzer{
+		indexRule,
+		indexRuleBinding,
+		metadataService,
+	}
+}
+
+func (a *Analyzer) BuildStreamSchema(ctx context.Context, metadata *commonv2.Metadata) (Schema, error) {
+	stream, err := a.metadataRepoImpl.Stream().Get(ctx, metadata)
+
+	if err != nil {
+		return nil, err
+	}
+
+	indexRules, err := a.metadataRepoImpl.IndexRules(context.TODO(), metadata)
+
+	if err != nil {
+		return nil, err
+	}
+
+	s := &schema{
+		stream:     stream,
+		indexRules: indexRules,
+		fieldMap:   make(map[string]*tagSpec),
+		entityList: stream.GetEntity().GetTagNames(),
+	}
+
+	// generate the schema of the fields for the traceSeries
+	for tagFamilyIdx, tagFamily := range stream.GetTagFamilies() {
+		for tagIdx, tagSpec := range tagFamily.GetTags() {
+			s.registerField(tagFamily.GetName(), tagSpec.GetName(), tagFamilyIdx, tagIdx, tagSpec)
+		}
+	}
+
+	return s, nil
+}
+
+func (a *Analyzer) Analyze(_ context.Context, criteria *streamv2.QueryRequest, metadata *commonv2.Metadata, s Schema) (Plan, error) {
+	// parse fields
+	plan, err := parseFields(criteria, metadata, s)
+	if err != nil {
+		return nil, err
+	}
+
+	if plan.Type() == PlanTraceIDFetch {
```

Comment:
A question here, since global index scan are much simplified, I would like to return unresolved plan here ASAP.

But the indexes to be used are determined after analysis with schema, how can we avoid unnecessary limit and offset here?

Reply:
Default Limit added to 100.